src/main/java/de/tum/in/www1/artemis/service/user/UserService.java

```diff
@@ -236,10 +240,26 @@ public User registerUser(UserDTO userDTO, String password) {
         Set<Authority> authorities = new HashSet<>();
         authorityRepository.findById(STUDENT.getAuthority()).ifPresent(authorities::add);
         newUser.setAuthorities(authorities);
-        saveUser(newUser);
+
         // we need to save first so that the user can be found in the database in the subsequent method
-        optionalVcsUserManagementService.ifPresent(vcsUserManagementService -> vcsUserManagementService.createVcsUser(newUser));
-        optionalCIUserManagementService.ifPresent(ciUserManagementService -> ciUserManagementService.createUser(newUser));
+        saveUser(newUser);
+
+        optionalVcsUserManagementService.ifPresent(vcsUserManagementService -> {
+            if (!vcsUserManagementService.canCreateVcsUser(newUser)) {
+                // Delete the user that was previously saved and throw
+                deleteUser(newUser);
```

Comment:
are you sure there can be no edge cases here where you actually would delete an actual user here?
What happens if an administrator creates an account manually between the registration and activation of a user account?

Reply:
This just happens during registration so I think it's safe. The user is first created in Artemis. If a VCS/CI exists, it checks if the user can be created there too. If that's not the case, we cancel registration by removing that user from Artemis and return an error.